### PR TITLE
enhancement(http source): emit Error and ErrorDropped events per instrumentation spec

### PR DESCRIPTION
**Pull request by neuronull**
Closes:  #14056.
Part of: #13995.

### CODE REVIEW

File: src/internal_events/http.rs [hunk 1]
```diff
@@ -129,3 +129,39 @@ impl<'a> InternalEvent for HttpDecompressError<'a> {
         counter!("parse_errors_total", 1);
     }
 }
+
+pub struct HttpInvalidRouteError {}
+
+impl InternalEvent for HttpInvalidRouteError {
+    fn emit(self) {
+        error!(
+            message = "Invalid route specified.",
```
Comment (spencergilbert) [inline src/internal_events/http.rs:138]:
```suggestion
            message = "Request received on unknown route.",
```
I don't like the term "specified" here, sounds like a config error.

File: src/sources/http.rs [hunk 2]
```diff
@@ -175,10 +175,12 @@ impl HttpSource for SimpleHttpSource {
                 }
                 Ok(None) => break,
                 Err(error) => {
+                    // Error is logged / emitted by `crate::codecs::Decoder`, no further
+                    // handling is needed here
```
Comment (spencergilbert) [inline src/sources/http.rs:179]: I was 99% sure I/we updated the internal events for codecs already - or was that just Encoding?

File: src/sources/util/http/prelude.rs [hunk 3]
```diff
@@ -82,6 +85,7 @@ pub trait HttpSource: Clone + Send + Sync + 'static {
                         Ok(())
                     } else {
                         debug!(message = "Path rejected.");
+                        emit!(HttpInvalidRouteError {});
                         Err(warp::reject::custom(ErrorMessage::new(
                             StatusCode::NOT_FOUND,
                             "Not found".to_string(),
```
Comment (spencergilbert) [inline src/sources/util/http/prelude.rs:91]: Re: my comment about the `HttpInvalidRouteError` - I imagine it should better align with the messages here. On that note, this `debug` seems duplicative of the new internal event.

File: src/sources/http.rs (hunk 2 above)
Comment (neuronull) [reply src/sources/http.rs:179]: It looks like you added them yes, and I made a modification in this PR:

https://github.com/vectordotdev/vector/pull/14021/files#diff-6f76e603a28a915a3113b9166df34b2e9872d285b28515226796b3997d5f88a2

Comment (spencergilbert) [reply src/sources/http.rs:179]: 😅 lol, yeah my brain is good after a week off.

File: src/sources/util/http/prelude.rs (hunk 3 above)
Comment (neuronull) [reply src/sources/util/http/prelude.rs:91]: Ah, good call. I've deleted the debug log and updated the Event in question.

File: src/internal_events/http.rs (hunk 1 above)
Comment (spencergilbert) [reply src/internal_events/http.rs:138]: (Feel free to tweak that, just first thought)

**Review by spencergilbert — APPROVED** [review]

**Pull request merged**